src/v/pandaproxy/server.h

```diff
@@ -182,8 +188,14 @@ class auth_ctx_server : public ctx_server<service_t> {
 
         template<std::invocable<kafka_client_cache&> Func>
         auto dispatch(Func&& func) {
-            return service().client_cache().invoke_on_cache(
-              user, std::forward<Func>(func));
+            // Access the cache on the appropriate shard.
+            return service().client_cache().invoke_on(
+              user_shard(user.name),
+              ss::smp_submit_to_options{context().smp_sg},
+              [func{std::forward<Func>(func)}](
+                kafka_client_cache& cache) mutable {
+                  return std::invoke(std::move(func), cache);
```

Comment:
Gate or lock here. OR really a gate/lock on the sharded service. The method `invoke_on_cache` was meant to the be the wrapper for the gate and lock.

Reply:
You don't need a gate; `kafka_client_cache` encapsulates a gate that protects async calls.
You don't need a lock; `kafka_client_cache` encapsulates a lock that protects against reentrancy of `clean_stale_clients`.

Reply:
Without the gate and lock, the test fails with 2k users (and 1k users but 2/10 runs) with a crash `rptest.services.utils.NodeCrash: <NodeCrash docker-rp-1: redpanda: /home/nyalia/redpanda/vbuild/debug/clang/rp_deps_install/include/seastar/core/iostream.hh:391: seastar::output_stream<char>::~output_stream() [CharType = char]: Assertion `!_in_batch && "Was this stream properly closed?"' failed.
`

There are also **many** connection errors due to timeout, likely because the broker crashed:
`41ea10>: Failed to establish a new connection: [Errno 111] Connection refused'))
    raise ConnectionError(e, request=request)
requests.exceptions.ConnectionError: HTTPConnectionPool(host='docker-rp-1', port=8082): Max retries exceeded with url: /topics (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7f4834380220>: Failed to establish a new connection: [Errno 111] Connection refused'))
`

The assert failure `Assertion `!_in_batch && "Was this stream properly closed?` indicates that a output_stream was destroyed/closed  while it still held bytes to transport.

Reply:
OK, so what's happening is that the client is being evicted during `do_connect`, and its lifetime isn't extended.

Try something like this:
```
diff --git a/src/v/kafka/client/client.cc b/src/v/kafka/client/client.cc
index 38abfdbd2..931963e46 100644
--- a/src/v/kafka/client/client.cc
+++ b/src/v/kafka/client/client.cc
@@ -63,9 +63,9 @@ ss::future<> client::do_connect(net::unresolved_address addr) {
     return make_broker(unknown_node_id, addr, _config)
       .then([this](shared_broker_t broker) {
           return broker->dispatch(metadata_request{.list_all_topics = true})
-            .then([this, broker](metadata_response res) {
-                return apply(std::move(res));
-            });
+            .then(
+              [this](metadata_response res) { return apply(std::move(res)); })
+            .finally([broker]() {});
       });
 }
```

Reply:
Done.
And this makes sense. I think Noah also mentioned something about this scenario but I thought I accounted for that already.

Reply:
https://github.com/redpanda-data/redpanda/pull/6781/commits/c08c4f767db8602bb6d13f3c5176d644c3003384